bwp.1
.nh
.TH BWP  1 2022-05-26 budlabs "User Manuals"
.SH NAME
.PP
bwp - budlabs wallpaper selector

.SH SYNOPSIS
.PP
.RS

.nf
bwp [--prev|-p --next|-n --random|-r] WALLPAPER
bwp --blur|-b [--prev|-p --next|-n --random|-r] WALLPAPER
bwp --lock|-l [--lock-options|-L OPTIONS] [--image-options|-I OPTIONS] [--wallpaper|-w --random|-r] WALLPAPER
bwp --delete|-d [--force|-f] WALLPAPER
bwp --add|-a    [--force|-f] FILE|DIR
bwp --rename|-x NEWNAME WALLPAPER
bwp --verbose|-V
bwp --help|-h
bwp --version|-v

.fi
.RE

.SH USAGE
.PP
The argument WALLPAPER is available for all
actions except \fB\fC--add\fR and can be either the name
of a file in \fB\fCBWP_DIR/walls\fR or the path to a
file (\fIif the file doesn't exist in the library,
it will get added before the action is applied\fP).
If the WALLPAPER argument is omitted and no
navigate action (\fB\fC--next\fR,\fB\fC--pre\fR,\fB\fC--random\fR) is
specified the action will get applied to
the \fBcurrentwall\fP (\fIthe file that
\fB\fCBWP_DIR/currentwall\fR links to.*\fP)

.PP
*except for the \fB\fC--lock\fR action that will use
\fB\fCBWP_DIR/currentlock\fR\&.

.SH OPTIONS
.PP
.RS

.nf
-a, --add                   | add FILE or files in DIRECTORY to the library.
-b, --blur                  | toggle/set blur.
-d, --delete                | delete WALLPAPER from library and history.
-f, --force                 | change behavior of --delete and --add options
-h, --help                  | print help and exit  
-I, --image-options OPTIONS | options passed to `convert` before locking the screen
-l, --lock                  | locks the computer with `i3lock(1)`.
-L, --lock-options  OPTIONS | Additional command line options passed to `i3lock`.
-n, --next                  | set next wallpaper according to BWP_DIR/history
-p, --prev                  | set previous wallpaper according to BWP_DIR/history
-r, --random                | get a random wallpaper from BWP_DIR/walls
-x, --rename        NEWNAME | rename WALLPAPER to NEWNAME
-V, --verbose               | louder output
-v, --version               | print version info and exit  
-w, --wallpaper             | use current wallpaper when --lock is used

.fi
.RE

.SS -a, --add
.PP
Use --force option to overwrite existing files.

.SS -b, --blur
.PP
If no WALLPAPER is specified the blur state of the
current wallpaper will get toggled.

.SS -d, --delete
.PP
Use \fB\fC--force\fR option to skip prompt.

.SS -I, --image-options OPTIONS
.PP
If OPTIONS is the path to a file, the content of
that file will be used. This options overrides
the environment variable:
\fB\fCBWP_LOCK_IMAGE_OPTIONS\fR

.SS -l, --lock
.PP
If no WALLPAPER is specified, the last image used
for lockscreen background will be used.

.SS -L, --lock-options  OPTIONS
.PP
If OPTIONS is the path to a file, the content of
that file will be used. This options overrides
the environment variable: \fB\fCBWP_LOCK_OPTIONS\fR

.SH ENVIRONMENT
.SS BWP_DIR
.PP
Directory where wallpapers and history file will get
stored.
defaults to: \fB\fC$HOME/.cache/wallpapers\fR

.SS BWP_GEOMETRY
.PP
($WIDTHx$HEIGHT) if set, images added to the library will
be resized (if needed) to the dimensions specified. If not
set, images will be resized (if needed) to the same
dimensions as the currently active monitor.
defaults to: \fB\fC""\fR

.SS BWP_COMMAND
.PP
Command used to set the wallpaper.
defaults to: \fB\fCfeh --no-fehbg --bg-fill\fR

.SS BWP_LOCK_OPTIONS
.PP
Additional command line options passed to \fB\fCi3lock\fR\&.
\fB\fC--lock-options\fR will override this variable if set.
defaults to: \fB\fC""\fR

.SS BWP_LOCK_IMAGE_OPTIONS
.PP
Additional command line options passed to \fB\fCconvert\fR before
locking the screen. \fB\fC--image-options\fR will override this
variable if set.
defaults to: \fB\fC""\fR

.SH CONTACT
.PP
Send bugs and feature requests to:
.br
https://github.com/budlabs/bwp/issues

.SH COPYRIGHT
.PP
Copyright (c) 2019-2022, budRich of budlabs
.br
SPDX-License-Identifier: MIT

.SH DEPENDENCIES
.PP
bash gawk sed ImageMagick
.br
i3lock (optional)

.SH SEE ALSO
.PP
\fB\fCi3lock(1)\fR, \fB\fCImageMagick(1)\fR, \fB\fCparallel(1)\fR, \fB\fCpngquant(1)\fR


bwp.sh
#!/usr/bin/env bash

: "${BWP_DIR:=$HOME/tmp/bwp}"
: "${BWP_GEOMETRY:=}"
: "${BWP_COMMAND:=feh --no-fehbg --bg-fill}"

main() {

  _wdir="$BWP_DIR/walls"
  _history="$BWP_DIR/history"

  [[ -d $_wdir ]] || {
    mkdir -p "$_wdir"
    mkdir -p "$BWP_DIR/blurs"
    mkdir -p "$BWP_DIR/locks"
    ERM "BWP_DIR didn't exist, created: $_wdir"
  }

  # bwp -a [-f] FILE|DIR  add to the library
  ((_o[a] == 1)) && {
    add_to_library "$*" > /dev/null
    exit
  }

  { [[ -n ${_o[x]} ]] || ((_o[d] == 1)) ;} \
    && ((_o[r]+_o[n]+_o[p] > 0)) \
    && ERH "-r -n -p is not allowed when deleting or renaming."

  _curpth="$(readlink "$BWP_DIR/currentwall")"
  _curnam="${_curpth##*/}"
  
  _wall="$(get_wall "$@")"

  if   ((_o[d] == 1)); then
    delete_wall "$_wall"
  elif [[ -n ${_o[x]} ]]; then
    rename_wall "${_o[x]}" "$_wall"
  elif ((_o[l] ==1)); then
    set_lock "$_wall"
  elif ((_o[w] == 1)) || ((_o[b] == 1)); then
    set_wall "$_wall"
  fi


}


__source="$(readlink -f "${BASH_SOURCE[0]}")"
__srcdir="${__source%/*}"
for f in "$__srcdir/lib/"* ; do . "$f" ; done


_help_msg="$(cat << 'EOB'
bwp - budlabs wallpaper manager

Usage
-----
bwp --wallpaper|-w [--prev|-p|--next|-n|--random|-r|WALLPAPER] set desktop wallpaper
bwp --blur|-b [--prev|-p|--next|-n|--random|-r|WALLPAPER]      toggle/set blur
bwp --lock|-l [--prev|-p|--next|-n|--random|-r|WALLPAPER]      lock screen
bwp --delete|-d [--force|-f] [WALLPAPER]                       delete wallpaper from library
bwp --add|-a [--force|-f] FILE|DIR                             add a new image to the library
bwp --rename|-x NEWNAME [WALLPAPER]                            rename WALLPAPER to NEWNAME
bwp --help|-h                                                  print this help
bwp --version|-v                                               print version info
EOB
)"

declare -A _o


# Execute getopt
ARGS=$(getopt --options "wblafx:rhvpnd" --longoptions "wallpaper,blur,lock,add,force,rename:,random,help,version,prev,next,delete" -- "$@");
 
#Bad arguments
if [ $? -ne 0 ]; then
  ERH no options, no action, no woman, no cry
fi
 
eval set -- "$ARGS"


while true; do
  case "$1" in
    -w|--wallpaper)
      echo walloption
      _o[w]=1
      shift
    ;;
    -b|--blur)
      _o[b]=1
      shift
    ;;
    -l|--lock)
      _o[l]=1
      shift
    ;;
    -a|--add)
      _o[a]=1
      shift
    ;;
    -f|--force)
      _o[f]=1
      shift
    ;;
    -x|--rename)
      shift
      _o[x]="$1"
      shift
    ;;
    -r|--random)
      echo random opt
      _o[r]=1
      shift
    ;;
    -h|--help)
      _o[h]=1
      shift
    ;;
    -v|--version)
      _o[v]=1
      shift
    ;;
    -p|--prev)
      _o[p]=1
      shift
    ;;
    -n|--next)
      _o[n]=1
      shift
    ;;
    -d|--delete)
      _o[d]=1
      shift
    ;;
    --)
      shift
      break
    ;;

  esac
done

unset f o co

main "${@}"
